File: src/main/java/com/ticketing/domain/concert/ConcertRepository.java
package com.ticketing.domain.concert;

import jakarta.persistence.LockModeType;
import jakarta.persistence.QueryHint;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Lock;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.jpa.repository.QueryHints;
import org.springframework.stereotype.Repository;

import java.util.Optional;

@Repository
public interface ConcertRepository extends JpaRepository<Concert, Long> {

    @Lock(LockModeType.PESSIMISTIC_WRITE)
    @QueryHints({
            @QueryHint(name = "jakarta.persistence.lock.timeout", value = "3000")
    })
    @Query("SELECT c FROM Concert c WHERE c.id = :id")
    Optional<Concert> findByIdWithLock(Long id);

    @Modifying
    @Query("UPDATE Concert c SET c.remainingSeats = :remainingSeats WHERE c.id = :concertId")
    int updateRemainingSeats(Long concertId, Integer remainingSeats);
}

File: src/main/kotlin/com/ticketing/domain/concert/ConcertRepository.kt
package com.ticketing.domain.concert

import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.stereotype.Repository
import java.time.LocalDateTime

@Repository
interface ConcertRepository : JpaRepository<Concert, Long> {
    fun findByStatus(status: ConcertStatus): List<Concert>
    fun findByStartDateTimeBetween(start: LocalDateTime, end: LocalDateTime): List<Concert>
}

